// @nxrocks/nx-spring-boot/src/generators/project/lib/create-library-files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLibraryFiles = createLibraryFiles;
const devkit_1 = require("@nx/devkit");
function createLibraryFiles(tree, options) {
    if (options.projectType === 'library') {
        devkit_1.logger.debug(`Generating sample files for library project...`);
        const substitutions = {
            language: options.language,
            projectName: options.projectName,
            packageName: options.packageName,
            ext: options.language === 'kotlin' ? 'kt' : options.language === 'groovy' ? 'groovy' : 'java',
            lineEnding: options.language === 'java' ? ';' : '',
            dot: '.',
            tpl: ''
        };
        const basePath = options.packageName?.replaceAll('.', '/');
        const mainPath = `${options.projectRoot}/src/main/${options.language}/${basePath}`;
        const testPath = `${options.projectRoot}/src/test/${options.language}/${basePath}`;
        (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/library/main'), mainPath, substitutions);
        (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/library/test'), testPath, substitutions);
    }
}
