// @nxrocks/nx-spring-boot/src/generators/project/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = normalizeOptions;
const project_name_and_root_utils_1 = require("@nx/devkit/src/generators/project-name-and-root-utils");
async function normalizeOptions(tree, options) {
    const { projectName, projectRoot, projectNameAndRootFormat, } = await (0, project_name_and_root_utils_1.determineProjectNameAndRootOptions)(tree, {
        name: options.name,
        projectType: options.projectType,
        directory: options.directory,
        projectNameAndRootFormat: options.projectNameAndRootFormat,
        //rootProject: options.rootProject,
        callingGenerator: '@nxrocks/nx-spring-boot:project',
    });
    options.projectNameAndRootFormat = projectNameAndRootFormat;
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    const projectDependencies = options.dependencies
        ?.split(',')
        .map((s) => s.trim())
        .filter((s) => !!s) || [];
    return {
        ...options,
        projectName,
        projectRoot,
        projectDependencies,
        parsedTags,
    };
}
